src/bootstrap/compile.rs

```diff
@@ -1163,6 +1163,33 @@ pub fn run_cargo(builder: &Builder, cargo: &mut Command, stamp: &Path, is_check:
 
             toplevel.push((file_stem, extension, expected_len));
         }
+        for filename in filenames {
+            // Skip files like executables
+            if !filename.ends_with(".rlib") &&
+               !filename.ends_with(".lib") &&
+               !is_dylib(&filename) &&
+               !(is_check && filename.ends_with(".rmeta")) {
+                break;
+            }
+
+            let filename = Path::new(&*filename);
+
+            // If this was an output file in the "host dir" we don't actually
+            // worry about it, it's not relevant for us.
```

Comment:
If the host dir contains a file that wasn't in the target dir, also promote it. This primarily is the case for things like proc macros, where if std depends on that, we'll need to make sure that the proc macro artifact is also promoted -- but since it's host only, it won't be in the target dir.